crates/db_schema/src/impls/tagline.rs

```diff
@@ -1,58 +1,70 @@
+  traits::Crud,
+  utils::{get_conn, limit_and_offset, DbPool},
 };
-use diesel::{insert_into, result::Error, ExpressionMethods, QueryDsl};
-use diesel_async::{AsyncPgConnection, RunQueryDsl};
+use diesel::{insert_into, result::Error, ExpressionMethods, OptionalExtension, QueryDsl};
+use diesel_async::RunQueryDsl;
 
-impl Tagline {
-  pub async fn replace(
-    pool: &mut DbPool<'_>,
-    for_local_site_id: LocalSiteId,
-    list_content: Option<Vec<String>>,
-  ) -> Result<Vec<Self>, Error> {
-    let conn = &mut get_conn(pool).await?;
-    if let Some(list) = list_content {
-      conn
-        .build_transaction()
-        .run(|conn| {
-          Box::pin(async move {
-            Self::clear(conn).await?;
+#[async_trait]
+impl Crud for Tagline {
+  type InsertForm = TaglineInsertForm;
+  type UpdateForm = TaglineUpdateForm;
+  type IdType = TaglineId;
 
-            for item in list {
-              let form = TaglineForm {
-                local_site_id: for_local_site_id,
-                content: item,
-                updated: None,
-              };
-              insert_into(tagline)
-                .values(form)
-                .get_result::<Self>(conn)
-                .await?;
-            }
-            Self::get_all(&mut conn.into(), for_local_site_id).await
-          }) as _
-        })
-        .await
-    } else {
-      Self::get_all(&mut conn.into(), for_local_site_id).await
-    }
+  async fn create(pool: &mut DbPool<'_>, form: &Self::InsertForm) -> Result<Self, Error> {
+    let conn = &mut get_conn(pool).await?;
+    insert_into(tagline)
+      .values(form)
+      .get_result::<Self>(conn)
+      .await
   }
 
-  async fn clear(conn: &mut AsyncPgConnection) -> Result<usize, Error> {
-    diesel::delete(tagline).execute(conn).await
+  async fn update(
+    pool: &mut DbPool<'_>,
+    tagline_id: TaglineId,
+    new_tagline: &Self::UpdateForm,
+  ) -> Result<Self, Error> {
+    let conn = &mut get_conn(pool).await?;
+    diesel::update(tagline.find(tagline_id))
+      .set(new_tagline)
+      .get_result::<Self>(conn)
+      .await
   }
+}
 
-  pub async fn get_all(
+impl Tagline {
+  pub async fn list(
     pool: &mut DbPool<'_>,
     for_local_site_id: LocalSiteId,
+    page: Option<i64>,
+    limit: Option<i64>,
   ) -> Result<Vec<Self>, Error> {
     let conn = &mut get_conn(pool).await?;
+    let (limit, offset) = limit_and_offset(page, limit)?;
     tagline
+      .order(published.desc())
+      .offset(offset)
+      .limit(limit)
       .filter(local_site_id.eq(for_local_site_id))
       .get_results::<Self>(conn)
       .await
   }
+
+  pub async fn get_random(
+    pool: &mut DbPool<'_>,
+    for_local_site_id: LocalSiteId,
+  ) -> Result<Option<Self>, Error> {
+    let conn = &mut get_conn(pool).await?;
+    sql_function!(fn random() -> Text);
+    tagline
+      .order(random())
+      .limit(1)
+      .filter(local_site_id.eq(for_local_site_id))
```

Comment:
Is there a potential edge case bug if filtering is done after the limit? Also, why match on a local site ID? Taglines should only be relevant in the context of the local site, so it's not like you have to worry about filtering out taglines from federated instances.

Reply:
`local_site_id` has been removed from the `Tagline` resource